docs/docs/validating-plugin-options.md

```diff
@@ -0,0 +1,224 @@
+   import { pluginOptionsSchema } from "../gatsby-node"
+
+   describe(`pluginOptionsSchema`, () => {
+     it(`should invalidate incorrect options`, () => {
+       const options = {
+         optionA: undefined, // Should be a boolean
+         message: 123, // Should be a string
+         optionB: `not a boolean`, // Should be a boolean
+       }
+       const { isValid, errors } = testPluginOptionsSchema(
+         pluginOptionsSchema,
+         options
+       )
+
+       expect(isValid).toBe(false)
+       expect(errors).toEqual([
+         `"optionA" is required`,
+         `"message" must be a string`,
+         `"optionB" must be a boolean`,
+       ])
+     })
+
+     it(`should validate correct options`, () => {
+       const options = {
+         optionA: false,
+         message: "string",
+         optionB: true,
+       }
+       const { isValid, errors } = testPluginOptionsSchema(
+         pluginOptionsSchema,
+         options
+       )
+
+       expect(isValid).toBe(true)
+       expect(errors).toEqual([])
+     })
+   })
+   ```
+
+## Best practices for option schemas
+
+The [Joi API documentation](https://joi.dev/api/) is a great reference to use while working on a `pluginOptionsSchema` as it shows all the available types and methods.
+
+Here are some specific Joi best practices for `pluginOptionsSchema`s:
+
+- [Add descriptions](#add-descriptions)
+- [Set default options](#set-default-options)
+- [Validate external access](#validate-external-access) where necessary
+- [Add custom error messages](#add-custom-error-messages) where useful
+- [Deprecate options](#deprecating-options) in a major version release rather than removing them
+
+### Add descriptions
+
+Make sure that every option and field has a `.description()` explaining its purpose. This is helpful for documentation as users can look at the schema and understand all the options. There might also be tooling in the future that auto-generates plugin option documentation from the schema.
+
+### Set default options
+
+You can use the `.default()` method to set a default value for an option. For example with `gatsby-plugin-console-log`, you can have the `message` option default to `"default message"` in case a user does not pass their own `message`:
+
+```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
+exports.pluginOptionsSchema = ({ Joi }) => {
+  return Joi.object({
+    optionA: Joi.boolean().required().description(`Enables optionA.`),
+    message: Joi.string()
+      .default(`default message`) // highlight-line
+      .description(`The message logged to the console.`),
+    optionB: Joi.boolean().description(`Enables optionB.`),
+  })
+}
+```
+
+Accessing `pluginOptions.message` would then log `"default message"` in all plugin APIs if the user does not supply their own value:
+
+```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
+exports.onPreInit = (_, pluginOptions) => {
```

Comment:
Sidebar question: Are plugin methods only available from onPreInit, or is it in any of those APIs?

> `exports.onPreInit = (_, pluginOptions)`

Also what is the `_` parameter syntax? I haven't seen that before and couldn't figure out how to google it.

Reply:
The pluginOptions are available in all of the APIs!

> Also what is the _ parameter syntax? I haven't seen that before and couldn't figure out how to google it.

This is a convention I thought was common to denote unused arguments. (`(_, pluginOptions)` === the first argument to the function isn't used, `(input, pluginOptions)` === it is used) However, that doesn't make any sense in documentation, I should just put the actual name there. Good call! 👍